|| [Installation](readme.md) ||  [Getting Started](gettingstarted.md) || [Tutorial](tutorial.md) || [Application](uses.md) || 

NEED TO UPDATE PICTURES WITH FINAL VERSION

# Understanding the Output
Here, you will find the output generated by the sample input data.

 - [Reference Folders](#references)
   - [References Used](#used)
   - [Readme](#readme)
   - [Summary](#summary)
 - [Read (Pair) Folders](#reads)
   - [Indexed Folders](#index)
   - [Combined Horizontal PDF](#horizontal)
   - [Combined Variation PDF](#variation)
   - [Phylip](#phylip)
   - [Variation Analysis](#analysis)
- [Other Folders](#other)
   - [Map Depth (All Refs)](#depth)
   - [Scaled Profiles (All Refs)](#profiles)

### Output Folder Contents
After a successful run, you will find a folder in your current directory or the directory specified when entering the command. This folder should be named starting with a time stamp and ending with '-RepeatProfiler'.

The time refers to the time you started the run and uses the timezone of your computer. If you are using Docker, the time stamp will be in UTC rather than the timezone of your local computer. 

##### Reference Folder<a name="references"></a>
Opening this folder, you will find the following file structure:

![](./pics/file_structure1.png)

Here, you will see individual folders for each of the references used in the run. Each reference folder contains all the analysis for that reference. These individual reference folders will be helpful in navigation if the pipeline is run with a multi-fasta sequence file. Your references are also kept in a single fasta file in the references_used<a name="used"></a> folder.

##### Readme <a name="readme"></a>
The ReadMe file contains a summary of the reads used, where they were located, and what index was assigned to each. The index is used in analyses and sample folder names. 

![](./pics/readme.png)

##### Summary<a name="summary"></a>
The_summary_final.csv table contains useful information about the length of the reference used, average coverage for reads, percent coverage, etc.

![](./pics/summary.png)

##### Read (Pair) Folders<a name="reads"></a>
Within each reference folder, there are individual folders for each read or pair of reads used in the run. These sample folders are labeled with the reference name and the index number<a name="index"></a>. The file structure looks like this:

![](./pics/file_structure2.png)

In these folders, you will find the real features of the tool. Prepare haha!

Each subfolder belongs to the read or read pair the reference sequence was run against. These folders contain the individual profiles produced by RepeatProfiler for the specific read(pair)-to-reference mapping. To learn more about these profiles, see the [tutorial](tutorial.md).

##### Scaled Profiles PDF<a name="horizontal"></a>
The scaled_profiles.pdf file contains all the profiles for this reference in a single file. This makes side-by-side comparisons easier and gives a nice overview of what to expect in the subfolders. Here is a portion of the file produced by this run:

![](./pics/combined_profile.png)

Each of these profiles represent the coverage of the read (pair) when mapped to the specfic reference (folder name). All the individual profiles are based on the same scale, so differences in color gradients are meaningful. By quickly looking at the profiles it is easy to tell which parts of the reference sequence have the highest (red) and lowest (blue) coverage. This can tell you a lot about this read and reference mapping including how and where the reference sequence is enriched in a species' reads compared to other species' reads. Color scaling accross different reads will help you in a great way to vizualize differences among very large sets of data.

##### Variant Profiles PDF<a name="variation"></a>
The variant_profiles.pdf file contains all the variation profiles for this reference in a single file. Again, this file was produced to enhance analytical inferences at a glance.

![](./pics/combined_variation.png)

##### Phylip<a name="phylip"></a>
There is much more information in the variation profiles than meets the eye. These plots were **produced from** the phylip file (found in the reference folder). This file is named as reference_name.phy. 

![](./pics/phylip.png)

For more information on the Phylip file, see the [tutorial](tutorial.md) page.

##### Variation Analysis<a name="analysis"></a>
The variation_analysis.tsv captures information from the plots. It contains the fraction of bases that didn't match for each read (pair) for this reference. Looking at this table, you can discover which plots have the most similar phylogentic positions and make hypotheses based on that.  

![](./pics/variation_analysis.png) 
 
### Other Folders
There are some miscellaneous folders produced. Here is an explanation of each.
 
##### Map Depth (All Refs)<a name="depth"></a>
This folder contains a csv file for each read-reference mapping which shows the depth mapped at each position of the reference.
 
![](./pics/mapdepth.png)
 
##### Scaled Profiles (All Refs)<a name="profiles"></a>
This folder contains the scaled profile pdfs for all references.

![](./pics/profiles.png)
